### DISCUSSION TITLE
Welcome to the LinkedIn DPH Framework

### DISCUSSION THREAD

**Discussion by mkanat** [linkedin/dph-framework#2] [Announcements]
Today we are open-sourcing the [LinkedIn Developer Productivity & Happiness Framework (DPH Framework)](https://linkedin.github.io/dph-framework/) - a collection of documents that describe the systems, processes, metrics, and feedback systems we use to understand our developers and their needs internally at LinkedIn. 

Now more than ever, developers are navigating so much change and new opportunity in this new era of Generative AI, so ensuring teams have the systems, processes, metrics and feedback systems to be successful is paramount. Our goal with this release was to offer an answer to one of the main questions we hear asked across the software industry, “How can I help my software development teams be more efficient, more effective, and happier?” We found that the best way to answer this question is through data, usually meaning metrics and feedback systems.

Over the last four years, LinkedIn has invested both time and resources to create a set of guidelines on how to design such a framework, starting from the high-level philosophy of how to think about the problem and getting down to specific best practices for implementers of data pipelines. We hope these materials can help you as you look to implement or improve productivity and happiness frameworks in your organizations, groups, or teams.

The Framework
The framework starts off by describing the high-level philosophy we use to guide our work, called [Goals, Signals, and Metrics](https://linkedin.github.io/dph-framework/goals-signals-metrics.html). This describes how to solve one of the most difficult problems in this space, “How do I choose what to measure?” We then use this framework to describe the specific goals and signals that we selected.

Next, we discuss our [Developer Personas](https://linkedin.github.io/dph-framework/developer-personas.html), a system we developed for categorizing developers into different groups based on their workflows, and thinking about priorities separately for each of those groups. The documentation provides information on how you can design your own Developer Personas system. It also describes our concept of [Persona Champions](https://linkedin.github.io/dph-framework/persona-champions.html), who are volunteers from across the engineering organization who understand the workflows and pain points for these engineers and can share those insights for consideration and incorporation into the system.

We’ve also included some important principles that can help you avoid developing a system that will require tons of rework. We discuss the difference between “data” and “insights,” when you want to use [qualitative (objective) data vs. qualitative (subjective) data](https://linkedin.github.io/dph-framework/qualitative-vs-quantitative.html), how to drive decisions (and provide the right data for your audience), and what data you should collect (including some thoughts about data schemas for engineering data).

When your data systems are in place, you’ll need to translate that data into concrete metrics, reports, and visualizations. We cover the key principles you’ll need to think about when designing any engineering-related metric and explain some of the common pitfalls in metric design. This includes understanding [the potential problems of aggregating multiple metrics into a “score,”](https://linkedin.github.io/dph-framework/scores.html) and the reasons why "volume of output" is not the best metric for judging the performance of individual software engineers.

To demonstrate how all of this works in practice, at the end we have a set of documents that detail a few specific metrics we have actually implemented at LinkedIn. It starts off by describing a few key concepts in developer productivity that are often thought about when we design metrics. We then provide [detailed definitions of a few example metrics](https://linkedin.github.io/dph-framework/example-metrics.html), covering both some company-wide developer productivity metrics, and a few metrics that are specific to an individual team. 

Finally, we round off the framework by explaining [why we chose those metrics](https://linkedin.github.io/dph-framework/why-our-metrics.html), with an in-depth answer for each metric in the set of example metrics. We also explain why we elected to not use some others.

Why Open-Source It?
We have made the DPH Framework open-source so that you can fork, modify, and re-use these documents however you wish inside of your own company, as long as you respect the Creative Commons license that is on the repository. 

We are also looking forward to community contributions that help move forward the state of the art in understanding software developers across the entire software industry. If there’s something missing in the documents that you’d like to see added, feel free to file an issue via our [GitHub Issues](https://github.com/linkedin/dph-framework/issues) tracker! If you just have questions or a discussion you’d like to have, participate in our [GitHub Discussions](https://github.com/linkedin/dph-framework/discussions).

And of course, if you want to contribute new text or improvements to the existing text, we welcome your contributions! Keep in mind that we hold this framework to a very high standard—we want it to be validated by real experience in high-level software development organizations, generally applicable across a wide range of software development environments, and that additional inputs are both interesting and accessible to a broader audience. If you think you have content that meets that bar and fits in with these documents, we would love to have your contribution! And honestly, even if you’re not sure, feel free to just send us a PR and find out!

We’re looking forward to what happens next!

**Comment by BlackFoxgamingstudio** [comment]
Why not create an actual class based, object based library for this? why just .md files around the high level philosophy? why can this framework not be explained to give code that lets a user track:

- personas
- Metrics
- Etc


Could this be something you let others develop via open opensources? in theory what guidance can your devs provide on the feasibility  of this by the community?  Bad idea?

**Reply by mkanat** [reply]
Excellent questions!

Our team does a lot of this work internally at LinkedIn. What we have found is:

1. The longest phase of development was in getting people to define and agree on their metrics, before coding even starts.
2. Once we did start, most of the development work was in creating data pipelines that translate data from custom data sources into coherent tables.
3. Even within the data pipeline work, _most_ of the engineering time goes to smoothing out anomalies in the data---specific problems that have to do with how your company works or how it has chosen to store data.
4. Once you have the data in place, what you need is a visualization tool. There are a ton of visualization tools out there in the world, including some open-source ones, and we didn't feel that the world would dramatically benefit from us doing the work to open-source any of ours.

So we decided that the best thing we could do was to open-source our guidance on _how to do all of this_, because the work seems so unique to every company. If some day there was a single, standard development platform that always stored its data in a consistent way across the entire developer lifecycle, and it was used by a significant percentage of the software industry, maybe that would change. But there is no such platform, at present. GitHub and Gitlab have some of that implemented, but not enough of it to matter for the huge diversity of different metrics that people want to measure.

**Reply by BlackFoxgamingstudio** [reply]
Thank you so much for the reply:) great work!

**Comment by valentinkravtchenko** [comment]
This is a fantastic initiative and it speaks volume about your passion for our fields' best practices.

As a project manager who went through a master on the topic, I would have enjoyed to possibility to consult your sources that supports the use of some concepts - since same words have different nuances.


If your team have metrics that can support the positive impact of our management tools (vs popular methodologies), it would be amazing to see a peer reviewed based white paper. 


In cases this is new to the reader, there is a great resources that is attempting something different but could serve a similar purpose : https://handbook.gitlab.com/


Again, fantastic initiative!

**Reply by mkanat** [reply]
Thanks, Valentin! Although there are many studies in this area that helped form our thinking, our sources for this are mostly our own experience implementing this and doing this work and similar experiences of our colleagues from across the industry.

In terms of metrics that show the impact, what sort of metrics are you thinking, and what aspect would you like to see data-based impact for?

Thanks for the link to the GitLab Handbook, I hadn't seen that and am looking forward to reading parts of it!

-Max

**Comment by ddelg4** [comment]
Thank you for combining 2 things I'm very passionate about - developer experience and open source:) Your documentation is meticulous and has given me many things to consider as well as additional clarity. 
One question - you focus on developer happiness and productivity, the intersection of which is focus time (or heads-down time). Have you explored this as a metric or signal? How to quantify that coding time, alot of which happens locally?
Thanks again!
~Diana

**Reply by mkanat** [reply]
Yeah, it's definitely something we have thought about. I know that other companies have implemented it. I believe Uber has a version of this metric, for example. At LinkedIn, we haven't yet implemented the metric internally, so we didn't feel confident enough to talk specifically about how it should be implemented.

You're right that counting coding time is hard. I think the compromise that most people take is looking at calendars, considering only events that have at least 2 people in them (because 1-person events could just be "I'm out of the office" or "this is my work time") and saying all other open time is available coding time. I believe there's more work required to exclude other types of calendar events that do represent work time, though.

-Max
